core/services/vrf/v2/coordinator_v2x_interface.go

```diff
@@ -0,0 +1,360 @@
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/log"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/vrf_coordinator_v2"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/vrf_coordinator_v2plus"
+	"github.com/smartcontractkit/chainlink/v2/core/services/vrf/vrfcommon"
+)
+
+// CoordinatorV2_X is an interface that allows us to use the same code for
+// both the V2 and V2.5 coordinators.
+type CoordinatorV2_X interface {
+	Address() common.Address
+	ParseRandomWordsRequested(log types.Log) (*RandomWordsRequested, error)
+	GetSubscription(opts *bind.CallOpts, subID uint64) (*GetSubscription, error)
+	GetConfig(opts *bind.CallOpts) (*GetConfig, error)
+	ParseLog(log types.Log) (generated.AbigenLog, error)
+	LogsWithTopics(keyHash common.Hash) map[common.Hash][][]log.Topic
+	Version() vrfcommon.Version
+}
+
+type RandomWordsRequested struct {
+	VRFVersion vrfcommon.Version
+	V2         *vrf_coordinator_v2.VRFCoordinatorV2RandomWordsRequested
+	V2Plus     *vrf_coordinator_v2plus.VRFCoordinatorV2PlusRandomWordsRequested
+}
+
+func (r *RandomWordsRequested) Raw() types.Log {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.Raw
+	}
+	return r.V2Plus.Raw
+}
+
+func (r *RandomWordsRequested) NumWords() uint32 {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.NumWords
+	}
+	return r.V2Plus.NumWords
+}
+
+func (r *RandomWordsRequested) SubID() uint64 {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.SubId
+	}
+	return r.V2Plus.SubId
+}
+
+func (r *RandomWordsRequested) MinimumRequestConfirmations() uint16 {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.MinimumRequestConfirmations
+	}
+	return r.V2Plus.MinimumRequestConfirmations
+}
+
+func (r *RandomWordsRequested) KeyHash() [32]byte {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.KeyHash
+	}
+	return r.V2Plus.KeyHash
+}
+
+func (r *RandomWordsRequested) RequestID() *big.Int {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.RequestId
+	}
+	return r.V2Plus.RequestId
+}
+
+func (r *RandomWordsRequested) PreSeed() *big.Int {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.PreSeed
+	}
+	return r.V2Plus.PreSeed
+}
+
+func (r *RandomWordsRequested) Sender() common.Address {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.Sender
+	}
+	return r.V2Plus.Sender
+}
+
+func (r *RandomWordsRequested) CallbackGasLimit() uint32 {
+	if r.VRFVersion == vrfcommon.V2 {
+		return r.V2.CallbackGasLimit
+	}
+	return r.V2Plus.CallbackGasLimit
+}
+
+func (r *RandomWordsRequested) NativePayment() bool {
+	if r.VRFVersion == vrfcommon.V2 {
+		return false
+	}
+	return r.V2Plus.NativePayment
+}
+
+type GetSubscription struct {
+	VRFVersion vrfcommon.Version
+	V2         vrf_coordinator_v2.GetSubscription
+	V2Plus     vrf_coordinator_v2plus.GetSubscription
+}
+
+func (s *GetSubscription) Balance() *big.Int {
```

Comment:
I worry that this pattern doesn't scale well, if we add another thing. You'd have to sweep through every getter and add a new condition.

Reply:
I would consider an implicit approach, i.e casting the v2/v2plus to an interface and storing it in the wrapper, and then for all of these getters calling  
```
val, ok := v2["Balance"].(*big.Int)
if !ok {
return nil, errors.New("...")
}
return val, nil
```

Reply:
I know you're not a huge fan of implicit though

Reply:
This'll only need to change if the subscription struct itself changes, which we should be extremely conservative in doing for V2+ upgrades.

Reply:
I mean for this entire `RandomWordsRequested` type. All the accessors here do an explicit `if s.VRFVersion == vrfcommon.V2` check that will be quite annoying to add to if we have a VRFCoordiantorPlusPlus or something.